executor/join.go

```diff
@@ -296,34 +315,104 @@ func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.C
-	if !sv.EnableRadixJoin {
-		return
+// partitionInnerRows re-order e.innerResults into sub-relations.
+func (e *HashJoinExec) partitionInnerRows() error {
+	if err := e.preAlloc4InnerParts(); err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	for i := 0; i < e.partConcurrency; i++ {
+		wg.Add(1)
+		workerID := i
+		go util.WithRecovery(func() {
+			e.doInnerPartition(workerID, wg)
+		}, nil)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) doInnerPartition(startIdx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	chkIdx, chkNum := startIdx, e.innerResult.NumChunks()
+	for ; chkIdx < chkNum; chkIdx += e.partConcurrency {
+		chk := e.innerResult.GetChunk(chkIdx)
+		for rowIdx, partPtr := range e.innerRowPrts[chkIdx] {
+			partIdx, rowPtr := partPtr.partitionIdx, partPtr.RowPtr
+			part := e.innerParts[partIdx]
+			part.Insert(rowPtr, chk.GetRow(rowIdx))
+		}
+	}
+}
+
+// preAlloc4InnerParts evaluates ptr4Partition and pre-alloc the memory space
+// for every inner row to help re-order the inner relation.
+// TODO: we need to evaluate the skewness for the partitions size, if the
+// skewness exceeds a threshold, we do not use partition phase.
+func (e *HashJoinExec) preAlloc4InnerParts() (err error) {
+	hasNull, keyBuf := false, make([]byte, 0, 64)
+	for chkIdx, chkNum := 0, e.innerResult.NumChunks(); chkIdx < chkNum; chkIdx++ {
+		chk := e.innerResult.GetChunk(chkIdx)
+		partPtrs := make([]ptr4Partition, chk.NumRows())
+		for rowIdx := 0; rowIdx < chk.NumRows(); rowIdx++ {
+			row := chk.GetRow(rowIdx)
+			hasNull, keyBuf, err = e.getJoinKeyFromChkRow(false, row, keyBuf)
+			if err != nil {
+				return err
+			}
+			if hasNull {
+				continue
+			}
+			joinHash := murmur3.Sum32(keyBuf)
+			partIdx := e.radixBits & joinHash
+			rowPtr := e.getPartition(partIdx).PreAlloc4Row(row)
+			partPtrs[rowIdx] = ptr4Partition{partIdx, rowPtr}
+		}
+		e.innerRowPrts = append(e.innerRowPrts, partPtrs)
+	}
+	return
+}
+
+func (e *HashJoinExec) getPartition(idx uint32) partition {
+	if e.innerParts[idx] == nil {
+		e.innerParts[idx] = chunk.NewList(e.innerExec.retTypes(), e.initCap, e.maxChunkSize)
 	}
+	return e.innerParts[idx]
+}
+
+// evalRadixBit evaluates the radix bit numbers.
+func (e *HashJoinExec) evalRadixBit() (needPartition bool) {
+	sv := e.ctx.GetSessionVars()
 	innerResultSize := float64(e.innerResult.GetMemTracker().BytesConsumed())
-	// To ensure that one partition of inner relation, one hash
-	// table and one partition of outer relation fit into the L2
-	// cache when the input data obeys the uniform distribution,
-	// we suppose every sub-partition of inner relation using
-	// three quarters of the L2 cache size.
+	// To ensure that one partition of inner relation, one hash table and one
+	// partition of outer relation fit into the L2 cache when the input data
+	// obeys the uniform distribution, we suppose every sub-partition of inner
+	// relation using three quarters of the L2 cache size.
```

Comment:
Maybe it's better to move this comment block in to the function comments. Because the function comments is far less than this comment.